segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/cache/DirectMemoryCache.java

```diff
@@ -0,0 +1,437 @@
+        return lastBlockAddress;
+    }
+
+    @Override
+    public int replace(int address, BufferView data) {
+        // The simplest and safest way to do a replace is to first insert the new data and then remove the old one. This
+        // will ensure the Cache remains consistent during the operation and we won't need to worry about concurrency or
+        // potential CacheFullExceptions.
+        int newAddress = insert(data);
+        delete(address);
+        return newAddress;
+    }
+
+    @Override
+    public int getAppendableLength(int currentLength) {
+        // If the current length is 0, we still allocate a block for it, so we need to indicate we can return the full
+        // block length for that. Otherwise we can only append up to the end of the last block.
+        int lastBlockLength = currentLength % this.layout.blockSize();
+        return currentLength == 0 ? this.layout.blockSize() : (lastBlockLength == 0 ? 0 : this.layout.blockSize() - lastBlockLength);
+    }
+
+    @Override
+    public int append(int address, int expectedLength, BufferView data) {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        Preconditions.checkArgument(address != CacheLayout.NO_ADDRESS, "Invalid address.");
+
+        // We can only append to fill the last block. For anything else a new write will be needed.
+        int appendedBytes = 0;
+        int expectedLastBlockLength = this.layout.blockSize() - getAppendableLength(expectedLength);
+        Preconditions.checkArgument(expectedLastBlockLength + data.getLength() <= this.layout.blockSize(),
+                "data is too long; use getAppendableLength() to determine how much data can be appended.");
+
+        int bufferId = this.layout.getBufferId(address);
+        int blockId = this.layout.getBlockId(address);
+        appendedBytes = this.buffers[bufferId].tryAppend(blockId, expectedLastBlockLength, data);
+
+        this.storedBytes.addAndGet(appendedBytes);
+        CacheMetrics.append(appendedBytes);
+        return appendedBytes;
+    }
+
+    @Override
+    public void delete(int address) {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        int deletedLength = 0;
+        while (address != CacheLayout.NO_ADDRESS) {
+            // Locate the Buffer-Block for the current address.
+            int bufferId = this.layout.getBufferId(address);
+            int blockId = this.layout.getBlockId(address);
+            DirectMemoryBuffer b = this.buffers[bufferId];
+
+            // Keep track if this was full before we removed anything from it.
+            boolean wasFull = !b.hasCapacity();
+
+            // Delete whatever we can from it and remember the successor.
+            DirectMemoryBuffer.DeleteResult result = b.delete(blockId);
+            address = result.getPredecessorAddress();
+            deletedLength += result.getDeletedLength();
+            if (wasFull) {
+                synchronized (this.availableBufferIds) {
```

Comment:
Shouldn't this be inside the if?

Reply:
This could go either way. 

The initial thinking here is that this needs to be synchronized with `getNextAvailableBuffer`. I prefer to have the buffer capacity check inside the synchronized block since otherwise it may get full again prior to adding it to the `availableBufferIds`.

However I now realize that I have code in `getNextAvailableBuffer` to handle full buffers and auto-remove them from there, so really this doesn't matter that much. I can make this change.